src/System.Net.Security/src/System/Net/SecureProtocols/_SecuritySafeHandles.Windows.cs

```diff
@@ -0,0 +1,1339 @@
+        {
+            return SetContextAttributes_SECURITY(phContext, contextAttribute, buffer);
+        }
+
+        private static int SetContextAttributes_SECURITY(
+            SafeDeleteContext phContext,
+            Interop.Secur32.ContextAttribute contextAttribute,
+            byte[] buffer)
+        {
+            int status = (int)Interop.SecurityStatus.InvalidHandle;
+            bool b = false;
+
+            // We don't want to be interrupted by thread abort exceptions or unexpected out-of-memory errors failing 
+            // to jit one of the following methods. So run within a CER non-interruptible block.
+            try
+            {
+                phContext.DangerousAddRef(ref b);
+            }
+            catch (Exception e)
+            {
+                if (b)
+                {
+                    phContext.DangerousRelease();
+                    b = false;
+                }
+                if (!(e is ObjectDisposedException))
+                    throw;
+            }
+            finally
+            {
+                if (b)
+                {
+                    status = Interop.Secur32.SetContextAttributesW(
+                        ref phContext._handle, contextAttribute, buffer, buffer.Length);
+                    phContext.DangerousRelease();
+                }
+            }
+
+            return status;
+        }
+    }
+
+    internal sealed class SafeFreeContextBuffer_SECURITY : SafeFreeContextBuffer
+    {
+        internal SafeFreeContextBuffer_SECURITY() : base() { }
+
+        override protected bool ReleaseHandle()
+        {
+            return Interop.Secur32.FreeContextBuffer(handle) == 0;
+        }
+    }
+
+    //
+    // Implementation of handles required CertFreeCertificateContext
+    //
+#if DEBUG
+    internal sealed class SafeFreeCertContext : DebugSafeHandle
+    {
+#else
+    internal sealed class SafeFreeCertContext : SafeHandleZeroOrMinusOneIsInvalid
+    {
+#endif
+
+        internal SafeFreeCertContext() : base(true) { }
+
+        // This must be ONLY called from this file within a CER.
+        internal unsafe void Set(IntPtr value)
+        {
+            this.handle = value;
+        }
+
+        private const uint CRYPT_ACQUIRE_SILENT_FLAG = 0x00000040;
+
+        override protected bool ReleaseHandle()
+        {
+            Interop.Crypt32.CertFreeCertificateContext(handle);
+            return true;
+        }
+    }
+
+    //
+    // Implementation of handles dependable on FreeCredentialsHandle
+    //
+#if DEBUG
+    internal abstract class SafeFreeCredentials : DebugSafeHandle
+    {
+#else
+    internal abstract class SafeFreeCredentials : SafeHandle
+    {
+#endif
+
+        internal Interop.Secur32.SSPIHandle _handle;    //should be always used as by ref in PInvokes parameters
+
+        protected SafeFreeCredentials() : base(IntPtr.Zero, true)
+        {
+            _handle = new Interop.Secur32.SSPIHandle();
+        }
+
+#if TRAVE
```

Comment:
I do think a different name would be good, and to standardize that across corefx.  My biggest issue with "TRAVE" is that it really looks like a typo, especially since v is next to c on a qwerty keyboard.